### PR TITLE
Mobile: add CDS builder APIs for all languages

### PR DESCRIPTION
**Pull request by caschoener**
Similar to https://github.com/envoyproxy/envoy/pull/25638, but for the CDS APIs. I'm using the enableCds boolean in all the different languages because it's possible for the cdsResourcesLocator to be empty, so we can't use that as a sentinel value.

### CODE REVIEW

File: mobile/library/swift/EngineBuilder.swift
Comment (abeyad) [inline mobile/library/swift/EngineBuilder.swift:625]: can this have a default value of "" for the cdsResourcesLocator?

File: mobile/test/java/io/envoyproxy/envoymobile/engine/EnvoyConfigurationTest.kt
Comment (abeyad) [inline mobile/test/java/io/envoyproxy/envoymobile/engine/EnvoyConfigurationTest.kt:361]: perhaps also a test case with no cdsResourcesLocator and timeout specified (and make sure they are the empty string and default timeout)

File: mobile/test/swift/EngineBuilderTests.swift
Comment (abeyad) [inline mobile/test/swift/EngineBuilderTests.swift:437]: nit: extra newline

File: mobile/test/swift/EngineBuilderTests.swift
Comment (abeyad) [inline mobile/test/swift/EngineBuilderTests.swift:436]: perhaps another test for default values for cdsResourcesLocator and timeout, similar to the suggestion for Kotlin

**Review by abeyad — DISMISSED** [review]

**Comment by jpsim** [comment]
Can we prevent these APIs from being reachable at compile time if they're missing build components?

Similar to how I wired up request compression, which is gated by a compile-time flag: https://github.com/envoyproxy/envoy/pull/25595

Whenever we can shift runtime errors to compile-time errors, we should do that to prevent issues like the one I fixed in https://github.com/envoyproxy/envoy/pull/26072

**Comment by caschoener** [comment]
How would I do that? It seems like if I put the builder API definitions inside an ifndef macro, then e.g. the jni_library layer would complain because it's looking for calls that don't exist.  

Or would I just need to have macros for including these functions in all the languages we supprt?

**Comment by caschoener** [comment]
Seems like yes for sure. I can do that

**Comment by caschoener** [comment]
Done. Let me know what you think!

File: mobile/docs/root/intro/version_history.rst
Comment (jpsim) [inline mobile/docs/root/intro/version_history.rst:23]: I continue to think that a separate nested xDS configuration object should be used so that we can enforce the expected grouping of values in the type system to make it impossible to miss one field that was required in a configuration. But not a blocker.

File: mobile/library/cc/engine_builder.cc
Comment (jpsim) [inline mobile/library/cc/engine_builder.cc:224]: Not a blocker, and probably out of scope for this PR, but ideally we'd have a dedicated `ENVOY_MOBILE_XDS` define / build flag that would force `ENVOY_GOOGLE_GRPC` to be enabled and allow decoupling EM's xDS support from google_grpc.

File: mobile/library/cc/engine_builder.cc
Comment (jpsim) [inline mobile/library/cc/engine_builder.cc:241]: Nice 👍

File: mobile/docs/root/intro/version_history.rst
Comment (jpsim) [reply mobile/docs/root/intro/version_history.rst:23]: Nit: missing issue

File: mobile/library/objective-c/EnvoyConfiguration.h
Comment (jpsim) [inline mobile/library/objective-c/EnvoyConfiguration.h:62]:
```suggestion
```

File: mobile/library/objective-c/EnvoyConfiguration.mm
Comment (jpsim) [inline mobile/library/objective-c/EnvoyConfiguration.mm:184]:
```suggestion
```

File: mobile/library/objective-c/EnvoyConfiguration.mm
Comment (jpsim) [inline mobile/library/objective-c/EnvoyConfiguration.mm:280]:
```suggestion
  if (self.cdsResourcesLocator != nil) {
```

File: mobile/library/swift/EngineBuilder.swift
Comment (jpsim) [inline mobile/library/swift/EngineBuilder.swift:631]:
```suggestion
  public func addCDSLayer(cdsResourcesLocator: String, timeoutSeconds: UInt32 = 0) -> Self {
    self.cdsResourcesLocator = cdsResourcesLocator
    self.cdsTimeoutSeconds = timeoutSeconds
    return self
  }
```

File: mobile/library/swift/EngineBuilder.swift
Comment (jpsim) [inline mobile/library/swift/EngineBuilder.swift:621]: Is this really just a name? In the docs, it's set as `xdstp://td-location.com/cluster_location`.

**Review by jpsim — APPROVED** [review]
Approving as to not block the merge considering I'm about to go on vacation, but hopefully some of my feedback can be considered before this ships.

File: mobile/docs/root/intro/version_history.rst
Comment (caschoener) [reply mobile/docs/root/intro/version_history.rst:23]: Resolving the nit, and the other point is a good idea but yeah maybe for another PR

File: mobile/library/swift/EngineBuilder.swift
Comment (abeyad) [reply mobile/library/swift/EngineBuilder.swift:621]: no, it's not a layer name, it is the xdstp resource URI for fetching Clusters

File: mobile/library/swift/EngineBuilder.swift
Comment (jpsim) [reply mobile/library/swift/EngineBuilder.swift:631]: > The idea being that it's legal to add a CDS config with no locator configured

Can you explain what the use case is for a CDS config with no locator?

Comment (caschoener) [reply mobile/library/swift/EngineBuilder.swift:631]: This is something @abeyad mentioned to me I think.  I'll leave it to him to verify or correct me on this.

Comment (caschoener) [reply mobile/library/swift/EngineBuilder.swift:631]: https://github.com/envoyproxy/envoy/blob/main/mobile/library/cc/engine_builder.h#L85 comment he wrote

File: mobile/library/swift/EngineBuilder.swift
Comment (jpsim) [inline mobile/library/swift/EngineBuilder.swift:622]: It would be useful to explain what a zero-second timeout does and what an empty locator does.

**Review by jpsim — APPROVED** [review]

File: mobile/library/swift/EngineBuilder.swift
Comment (abeyad) [reply mobile/library/swift/EngineBuilder.swift:631]: that's correct, cds resources locator is only for xdstp namespaces, for non-xdstp, nothing needs to be specified and a wildcard is inferred.

**Review by jpsim — DISMISSED** [review]

**Review by jpsim — APPROVED** [review]

**Review by jpsim — APPROVED** [review]

**Pull request merged**
